### ISSUE TITLE
User Authentication with Alt, React-Router, and auth tokens

### ISSUE THREAD

**Issue by c0d3ster** [njtrettel/SolverCAD#3]
After looking through numerous examples on how projects are authenticating users, it looks as if our best bet is to use hashed authentication tokens. [JWT authentication tokens](https://jwt.io/introduction/) are a nice, secure way to manage our current user state.   We could figure out how to implement these ourselves, but I've yet to find an example that is clear enough to get me started on that and the people that made JWT tokens provide this library/service: [Auth0](https://auth0.com/why-auth0) - free account includes 7000 users but set up seems a bit complicated and overkill for what we want to do
The [Airpal app](https://github.com/airbnb/airpal/tree/master/src/main/resources/assets/javascripts) by AirBnB uses custom auth tokens and xhr requests, but the way the app works is confusing to me.

Basically, I'm just trying to identify the fastest and easiest way I can start moving forward with Authenticating various pages to make them only accessible when the user is logged in.   Another useful example on the Authentication for the React-Router side of things can be found [HERE](https://github.com/ReactTraining/react-router/tree/latest/examples/auth-with-shared-root).  That example is very similar to what we need, in that it loads a home page that varies based on user status and has some pages requiring the user to be logged in (and redirects to the home/signin page if you are not logged in).  The only thing is this is not using Alt (it's just using an auth.js file and localStorage), so I've yet to bridge the gap to making that work with our Flux architecture.

**Comment by njtrettel** [comment]
How should we handle authentication with the DB? MongoDB has [authentication methods](https://docs.mongodb.com/manual/core/authentication/) as well. 

We could authenticate the user on the client via JWT and only allow CRUD operations after authentication. Then we could limit MongoDB connections to localhost only, with no user authentication. 

The other option is to authenticate the user on the client by GET/POST to an endpoint on the server, which would check with MongoDB, and respond either success or failure. This response would be used to set currUser or throw some type of auth error.

**Comment by njtrettel** [comment]
Current strategy is to use JWT to prevent checking the DB for every authentication. The user will authenticate themselves once and store a returned token (it expires after _x_ time) with all necessary user information. The DB will store a hashed password using `bcrypt` on the server side. 

The password, however, will not be hashed until it gets to the server. And obviously sending plain text passwords over `http` is not a good idea. The solution is to use `https` since it is an encrypted connection. We are most likely going to switch the server to run on `https` instead. I cannot find a single website that still uses `http://`.
